### ISSUE TITLE
 0 bytes read. Closing.

### ISSUE THREAD

**Issue by ThePaceMaker** [statianzo/Fleck#231]
I'm running IoT devices connected to Fleck via SSL. It sends a message every minute with a ping/pong occurring every 30 seconds. All JSON packets are the same in structure except for floating points that change from the sensor readings.

Intermittently I have Fleck disconnecting the IoT devices with "0 bytes read. Closing." message.  The task returns -1: 

`Task<int> Receive(byte[] buffer, Action<int> callback, Action<Exception> error, int offset = 0);`

In the Fleck log file, I get the following message returned with no further debug information: 

`DEBUG Fleck.FleckLog - 0 bytes read. Closing.`

I have tried to put a break point in and follow the stack trace back up the tree, to see what message is coming in, when that method is hit... but it only takes me so far back.

My gut feeling tells me that it's the message that is blowing up the connection. 

Long story short... any pointers how I could debug this or if someone has had a similar problem, to point me in the right direction?

Thanks in advance!

**Comment by ThePaceMaker** [comment]
So the following exception is being thrown causing the issue above: 

**EndRead can only be called once for each asynchronous operation.**

`at System.Net.Security._SslStream.EndRead(IAsyncResult asyncResult)
at System.Net.Security.SslStream.EndRead(IAsyncResult asyncResult)
at Fleck.QueuedStream.EndRead(IAsyncResult asyncResult) in QueuedStream.cs:line 93`

**The actual code block**

`
 public override int EndRead(IAsyncResult asyncResult) {
try {
   return _stream.EndRead(asyncResult);
  } catch (Exception) {
   return -1;
  }
 }
`

My initial assumption was that the message was blowing up the stream due to the fact that it happens intermittently... but that doesn't seem to be the case. 

Has anyone out interweb had an issue like this?

**Comment by ThePaceMaker** [comment]
There seems to be a problem with the QueuedStream object.  I have run out of time trying to figure out how to fix it. I will come back at a later stage to it to see if I can debug it and maybe implement a fix.

In the meantime, as a "quick fix", I have nginx reverse proxy the requests, terminating the SSL there and then connecting through to Fleck via normal sockets.  

Seems to be working at the moment.
